chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,181 @@
+        elif recv_rank == comm.rank:
+            # I am the receiver
+            assert send_rank != comm.rank
+            recv_data = comm.recv_obj(source=send_rank, tag=0)
+            data += recv_data
+
+    return data
+
+
+def _adjust_data_length(comm, data):
+    # The process which has maximum number of elements sends its data
+    # to other processes
+
+    length_all = [x[0] for x in comm.allgather(np.array([len(data)]))]
+    length_final = max(length_all)
+
+    root_rank = next(r for r in range(comm.size)
+                     if length_all[r] == length_final)
+
+    if comm.rank == root_rank:
+        offset = 0
+        for dest_rank in range(comm.size):
+            if comm.rank == dest_rank:
+                continue
+
+            send_cnt = length_final - length_all[dest_rank]
+            if send_cnt > 0:
+                comm.send_obj(data[offset:offset + send_cnt], dest_rank, tag=0)
+                offset += send_cnt
+    else:
+        if len(data) < length_final:
+            data += comm.recv_obj(source=root_rank, tag=0)
+        assert len(data) == length_final
+
+
+def shuffle_data_chunks(comm, data_chunks, force_equal_length=True,
+                        chunk_size=10000):
+    """Exchange chunks of data between all processes
+
+    This function is useful when `scatter_dataset` is not suitable.
+    For instance, the data is huge or the total length is unknown.
+
+    :param comm: ChainerMN communicator
+    :param data_chunks: Sequence or generator to read chunks.
+    :param force_equal_length: Whether data length of each process is adjusted
+                               by copying some elements, so that all processes
+                               have exactly the same length of data.
+                               This is required in training for correct
+                               iteration/epoch counting.
+                               In evaluation, however, the option can be False
+                               if you don't want duplicated elements.
+    :param chunk_size: Number of elements read from `data_chunks` at a time
+    :return: Shuffled data (in a list)
+    """
+
+    if chunk_size <= 1:
+        raise ValueError()
+
+    if force_equal_length not in [None, True, False]:
+        raise ValueError('Wrong value for `force_equal_length`:'
+                         ' {}'.format(force_equal_length))
+
+    data_chunks = iter(data_chunks)
+    data = []
+
+    # repeat until all processes consume all data
+    while True:
+        # Read a chunk of data; we need to use itertools.islice
+        # to support both of list-like objects and generators
+        chunk = list(itertools.islice(data_chunks, chunk_size))
+
+        if chunk is None:
+            length = np.array([0])
+        else:
+            length = np.array([len(chunk)])
+
+        # How many elements does each process have?
+        chunk_length_all = [x[0] for x in comm.allgather(length)]
```

Comment:
`chunk_lengths` or just `lengths` would be better variable name.

Reply:
fixed.